### DISCUSSION TITLE
What's the best way to handle async validation of a group of fields?

### DISCUSSION THREAD

**Discussion by jannisg** [react-hook-form/react-hook-form#3743] [Q&A] [answered]
I'm building a customer signup form which has a bunch of the usual name, email etc fields but also 6 fields —name(s), date of birth and driver license fields (2 fields)— which should be validated as a group basically.

My initial idea was to use `watch([...myLinkedFields])` and then trigger a `useEffect` which internally uses `setError`.
This mostly worked however it became a bit of an issue during the submit -> reset -> rerender cycle where the form would rerender and instantly validate showing the license validation API errors again on page refresh with persisted values.

I then tried building an async validator so it's not triggered as a `watch` + `useEffect` combo so it would be part of the same validation cycle when the form renders/triggers however I quickly realised that there was no good way (that I could see) that would allow me to access the other fields from within the async validator on the individual field level while still only running the logic if my own validator outcome is true.

The example I came up with (lacks debouncing wrappers for brevity):

```js
const isNotSelf = (self) => item => item !== self;
const isTrue = (value) => value === true;

const identityCheckValidator = (name: string) => async (value: string) => {
  const requiredFields = ["firstName", "lastName", "dateOfBirth", "licenseNo", "licenseVersion"];

  // Check all "other" fields are valid
  const otherFieldsValid = await Promise.all(
    requiredFields
      .filter(isNotSelf(name))
      .map(async (fieldName) => {
        const value = getValues(fieldName);
        const isFieldValid = await trigger(fieldName);

        return !!value && isFieldValid;
      })
  ).then((results: boolean[]) => results.every(isTrue));

  /**
   * Issue:
   *  I cannot call `await trigger(name)` (self validation) to run my other
   *  validators to assert that the current field is actually valid itself.
   *  If I did call that, it would result in an infinite loop since the
   *  `await trigger(name)` would re-fire this async validator function.
   */
  if (otherFieldsValid) {
    const {
      invalidFields,
      missingFields
    }: { invalidFields?: string[], missingFields?: string[] } = await myApiCall(getValues(requiredFields));

    // Show error messages for other fields if the API complained about the value
    if (missingFields) {
      const message = "This field is required.";

      missingFields
        .filter(isNotSelf(name))
        .forEach((fieldName: string) => {
          setError(fieldName, { type: "identityCheck", message });
        });

      if (missingFields.includes(name)) {
        return message; // return a string for the local field that triggered this function.
      }
    }

    if (invalidFields) {
      const message = "This field is invalid.";

      invalidFields
        .filter(isNotSelf(name))
        .forEach((fieldName: string) => {
          setError(fieldName, { type: "identityCheck", message });
        });

      if (invalidFields.includes(name)) {
        return message; // return a string for the local field that triggered this function.
      }
    }

    return true; // identity check presumably passed, no errors.
  }
};
```

I would then basically build a register like this:
```jsx
<input
    name="lastName"
    ref={
      register({
        required: "This field is required",
        minLength: {
          value: 2,
          message: "Must be more than 1 character"
        },
        validate: {
          name: value => /^[A-Za-z]$/i.test(value) || "A valid name is required",
          identityCheck: identityCheckValidator("lastName")
        }
      })
    }
  />
```

### Questions:
1. Within my async validator, how do I test that the other validation rules have passed? (`required`, `minLength` and `validate.name`)
2. Is there a better way to infer the current field name within the validator so I don't need to pass it in and repeat the string "lastName" in the example above?

This might overcomplicate things but what if the validator function instead of `(value) => boolean | string` was actually:

```js
const validator = (value, meta) => {
  const {
    field: { name }, // perhaps all props on the field?
    rules: {
      // same as per the `rules` definition?
    },
  } = meta;

  // I could now build logic without having to repeat the field's "name" string
  // AND I could (while not ideal) run other validation rules inside this validator
}
```

A nice to have would be if the `rules` could tell me which one the active function is so I can basically run the others manually but not sure how that would work to be honest.

---

Anyway, I have a working solution for now but felt that maybe this had a better solution or perhaps would be worth documenting (if there is a single "best" solution here) in the FAQ maybe?

**Comment by bluebill1049** [comment, marked as answer]
Thanks for the detailed question @jannisg good context as well, if there are many fields that need aysnc validation as a group, maybe `resolver` is the better solution? because you have access to the entire form value and you can probably have better custom logic inside the `resolver`as well.

**Reply by jannisg** [reply]
Thanks for the reply. I actually hadn't looked into `resolver` much before, very interesting indeed!

It does seem like a great place to put this logic since I could first run a base level `validationSchema` call (as per the "custom" example) that checks that generally those values adhere to some sensible inputs/formats, and then once I am passed that stage, I could wire up my API call to get those API errors back before returning the `{ values, errors }` objects.

One thing that —_as someone that hasn't used these resolvers before_— isn't quite clear to me is how a resolver interacts with `onBlur` and `onChange` `mode`s of the `useForm`?

More specifically, it seems that the argument to the resolver function is `values` and that validation occurs for all fields at once.
If that is the case, how do I ensure that errors are not appearing before a field has been blurred if my form is in `mode: "onBlur"`?

**Reply by bluebill1049** [reply]
"mode" should take care when the validation should be fired, in the next major version we are looking into provide input name to the resolver as well, so users have full control with resolver just as build-in validators.

**Reply by jannisg** [reply]
That sounds great, thank you for the answers :)
I will give a custom resolver a go tomorrow morning.

**Reply by bluebill1049** [reply]
pleasure and happy to help if you stuck on an issues with resolver.
